// @xviz/schema/dist/es5/proto-struct-wrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StructEncode = exports.StructDecode = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _protobufjs = require("@xviz/protobufjs");

var is = _interopRequireWildcard(require("is"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var StructEncode = function () {
  function StructEncode(options) {
    (0, _classCallCheck2["default"])(this, StructEncode);
    options = options || {};
    this.seenObjects = new Set();
    this.removeCircular = options.removeCircular === true;
    this.stringify = options.stringify === true;
  }

  (0, _createClass2["default"])(StructEncode, [{
    key: "encodeStruct",
    value: function encodeStruct(obj) {
      var convertedObject = {
        fields: {}
      };
      this.seenObjects.add(obj);

      for (var prop in obj) {
        if (obj.hasOwnProperty(prop)) {
          var value = obj[prop];

          if (is.undefined(value)) {
            continue;
          }

          convertedObject.fields[prop] = this.encodeValue(value);
        }
      }

      this.seenObjects["delete"](obj);
      return convertedObject;
    }
  }, {
    key: "encodeValue",
    value: function encodeValue(value) {
      var convertedValue;

      if (is["null"](value)) {
        convertedValue = {
          nullValue: 0
        };
      } else if (is.number(value)) {
        convertedValue = {
          numberValue: value
        };
      } else if (is.string(value)) {
        convertedValue = {
          stringValue: value
        };
      } else if (is["boolean"](value)) {
        convertedValue = {
          boolValue: value
        };
      } else if (is.object(value)) {
        if (this.seenObjects.has(value)) {
          if (!this.removeCircular) {
            throw new Error(['This object contains a circular reference. To automatically', 'remove it, set the `removeCircular` option to true.'].join(' '));
          }

          convertedValue = {
            stringValue: '[Circular]'
          };
        } else {
          convertedValue = {
            structValue: this.encodeStruct(value)
          };
        }
      } else if (is.array(value)) {
        convertedValue = {
          listValue: {
            values: value.map(this.encodeValue.bind(this))
          }
        };
      } else {
        if (!this.stringify) {
          throw new Error("Value of type ".concat((0, _typeof2["default"])(value), " not recognized."));
        }

        convertedValue = {
          stringValue: String(value)
        };
      }

      return convertedValue;
    }
  }]);
  return StructEncode;
}();

exports.StructEncode = StructEncode;

var StructDecode = function () {
  function StructDecode() {
    (0, _classCallCheck2["default"])(this, StructDecode);
  }

  (0, _createClass2["default"])(StructDecode, null, [{
    key: "decodeValue",
    value: function decodeValue(value) {
      switch (value.kind) {
        case 'structValue':
          {
            return StructDecode.decodeStruct(value.structValue);
          }

        case 'nullValue':
          {
            return null;
          }

        case 'listValue':
          {
            return value.listValue.values.map(StructDecode.decodeValue);
          }

        default:
          {
            return value[value.kind];
          }
      }
    }
  }, {
    key: "decodeStruct",
    value: function decodeStruct(struct) {
      var convertedObject = {};

      for (var prop in struct.fields) {
        if (struct.fields.hasOwnProperty(prop)) {
          var value = struct.fields[prop];
          convertedObject[prop] = StructDecode.decodeValue(value);
        }
      }

      return convertedObject;
    }
  }]);
  return StructDecode;
}();

exports.StructDecode = StructDecode;
_protobufjs.wrappers['.google.protobuf.Value'] = {
  fromObject: function fromObject(object) {
    if (object) {
      return new StructEncode().encodeValue(object);
    }

    return this.fromObject(object);
  },
  toObject: function toObject(message) {
    return StructDecode.decodeValue(message);
  }
};
_protobufjs.wrappers['.google.protobuf.Struct'] = {
  fromObject: function fromObject(object) {
    if (object) {
      return new StructEncode().encodeStruct(object);
    }

    return this.fromObject(object);
  },
  toObject: function toObject(message) {
    return StructDecode.decodeStruct(message);
  }
};
